Add an option section to API reference page (#407)
PostgREST/postgrest-docs: api.rst

  The OpenAPI information can go out of date as the schema changes under a running server. To learn how to refresh the cache see :ref:`schema_reloading`.

.. _options_requests:

OPTIONS
=======

You can verify which HTTP methods are allowed on endpoints for tables and views using an OPTIONS request. These methods are allowed depending on what operations *can* be done on the table or view, not on the database permissions assigned to them.

CORS
----

> PostgREST sets highly permissive cross origin resource sharing, that is why it accepts Ajax requests from any domain.

Reviewer: Now that I gave it a second look. I think this should go last on the OPTIONS section - otherwise it breaks the flow of the text.

> Tables
> ------

Reviewer: One issue with this heading, its URL fragment is simply "tables":
https://postgrest--407.org.readthedocs.build/en/407/api.html#tables

That might conflict with a future fragment. Maybe rename it to `options-tables`:
```suggestion

.. _options_tables:

Tables
------
```

Suppose you make this API request to an existing table named ``people``.

.. code-block:: http

  OPTIONS /people HTTP/1.1

Then, the response you get is:

.. code-block:: http

  HTTP/1.1 200 OK
  Allow: OPTIONS,GET,HEAD,POST,PUT,PATCH,DELETE

> 
> Views
> -----

Reviewer: Ditto:
```suggestion

.. _options_views:

Views
-----
```

> The methods that can be used on views are determined by the presence of INSTEAD OF TRIGGERS and primary key columns:

Reviewer: I think we can leave the exception on the table row.

```suggestion
The methods that can be used on views are determined by the presence of INSTEAD OF TRIGGERS:
```

.. table::
   :widths: auto

   +--------------------+-------------------------------------------------------------------------------------------------+
   | Method allowed     | View's requirements                                                                             |
   +====================+=================================================================================================+
   | OPTIONS, GET, HEAD | None (Always allowed)                                                                           |
   +--------------------+-------------------------------------------------------------------------------------------------+
   | POST               | INSTEAD OF INSERT TRIGGER                                                                       |
   +--------------------+-------------------------------------------------------------------------------------------------+

>    | PUT                | INSTEAD OF INSERT TRIGGER, INSTEAD OF UPDATE TRIGGER, primary key columns                       |

Reviewer: ```suggestion
   | PUT                | INSTEAD OF INSERT TRIGGER, INSTEAD OF UPDATE TRIGGER, also requires the presence of a  primary key                       |
```

   +--------------------+-------------------------------------------------------------------------------------------------+
   | PATCH              | INSTEAD OF UPDATE TRIGGER                                                                       |
   +--------------------+-------------------------------------------------------------------------------------------------+
   | DELETE             | INSTEAD OF DELETE TRIGGER                                                                       |
   +--------------------+-------------------------------------------------------------------------------------------------+
   | All the above methods are allowed for                                                                                |
   | `auto-updatable views <https://www.postgresql.org/docs/current/sql-createview.html#SQL-CREATEVIEW-UPDATABLE-VIEWS>`_ |
   +--------------------+-------------------------------------------------------------------------------------------------+

Functions
~~~~~~~~~

For database function endpoints, OPTIONS requests are not supported.

.. important::

>   Whenever you add or remove tables or views, or modify a view's definition, primary key columns or INSTEAD OF TRIGGERS on the database, you must refresh PostgREST's schema cache for OPTIONS requests to work properly. See the section :ref:`schema_reloading`.

Reviewer: Summarizing this a bit, since it's mostly about views
```suggestion
  Whenever you modify a view's INSTEAD OF TRIGGERS on the database, you must refresh PostgREST's schema cache for OPTIONS requests to work properly. See the section :ref:`schema_reloading`.
```

Author: > Summarizing this a bit, since it's mostly about views

I think at least a mention to table creation should be kept there. If a new table is created the OPTIONS result will be a 404 till the schema is reloaded, for instance.

Reviewer: Cool, I agree. A mention to tables' existence should be kept.

  The OpenAPI information can go out of date as the schema changes under a running server. To learn how to refresh the cache see :ref:`schema_reloading`.

.. _options_requests:

OPTIONS
=======

> You can verify which HTTP methods are allowed on endpoints for tables and views using an OPTIONS request. These methods are allowed depending on what operations *can* be done on the table or view, not on the database permissions assigned to them.

Reviewer: ```suggestion
You can verify which HTTP methods are allowed on endpoints for tables and views by using an OPTIONS request. These methods are allowed depending on what operations *can* be done on the table or view, not on the database permissions assigned to them.
```

.. _multiple-schemas:

> OPTIONS requests

Reviewer: I think just "OPTIONS" would do.
```suggestion
OPTIONS
```

otherwise, the methods that can be used are determined by the view's INSTEAD OF TRIGGERS and primary key columns:

=================== ==========================================================================
 Method allowed      View's requirements
=================== ==========================================================================
 OPTIONS, GET, HEAD  None (Always allowed)
 POST                INSTEAD OF INSERT TRIGGER
 PUT                 INSTEAD OF INSERT TRIGGER, INSTEAD OF UPDATE TRIGGER, primary key columns
 PATH                INSTEAD OF UPDATE TRIGGER
 DELETE              INSTEAD OF DELETE TRIGGER
=================== ==========================================================================

For instance, suppose you have a **non** `auto-updatable` view named ``people_view`` that has an INSTEAD OF INSERT and an INSTEAD OF DELETE trigger.
When you make this request:

.. code-block:: http

  OPTIONS /people_view HTTP/1.1

The response you get is:

.. code-block:: http

  HTTP/1.1 200 OK
  Allow: OPTIONS,GET,HEAD,POST,DELETE

Functions
~~~~~~~~~

> For database functions endpoints, OPTIONS requests are not supported.

Reviewer: Good to be explicit about this :+1:
